src/wp-includes/global-styles-and-settings.php

```diff
@@ -306,8 +306,47 @@ function wp_add_global_styles_for_blocks() {
 
 	$tree        = WP_Theme_JSON_Resolver::get_merged_data();
 	$block_nodes = $tree->get_styles_block_nodes();
+
+	$can_use_cached = ! wp_is_development_mode( 'theme' );
+	if ( $can_use_cached ) {
+		// md5 is a costly operation, so we hashing global settings and block_node in a single call.
+		$hash = md5(
+			wp_json_encode(
+				array(
+					'global_setting' => wp_get_global_settings(),
+					'block_nodes'    => $block_nodes,
+				)
+			)
+		);
+
+		$cache_key = "wp_styles_for_blocks:$hash";
+		$cached    = get_site_transient( $cache_key );
+		if ( ! is_array( $cached ) ) {
+			$cached = array();
+		}
+	}
+
+	$update_cache = false;
+
 	foreach ( $block_nodes as $metadata ) {
-		$block_css = $tree->get_styles_for_block( $metadata );
+		if ( isset( $metadata['name'] ) ) {
+			$block_name = $metadata['name'];
+		} elseif ( ! empty( $metadata['path'] ) ) {
+			// The likes of block element styles from theme.json do not have  $metadata['name'] set.
+			$block_name = wp_get_block_name_from_theme_json_path( $metadata['path'] );
+		}
+
+		if ( $can_use_cached && $block_name ) {
+			if ( isset( $cached[ $block_name ] ) ) {
```

Comment:
@joemcgill  I remove `str_replace( '\', '-'`. I felt it was unnecessary overhead, given that we only need a valid string here, which we get from `$block_name` itself.